tests/koji/test_koji_vmi.py

```diff
@@ -238,3 +239,140 @@ def test_koji_vmi_compound_product_name(fake_koji, koji_dir):
         signing_key=None,
         release=rel_obj,
     )
+
+
+def test_coreos_assembler_image(fake_koji, koji_dir):
+    archives = []
+    images = {
+        "meta.json": "json",
+        "commitmeta.json": "json",
+        "coreos-assembler-git.tar.gz": "tar",
+    }
+    for file_name, type_name in images.items():
+        archives.append(
+            {
+                "btype": "image",
+                "btype_id": 4,
+                "build_id": 123456,
+                "buildroot_id": 1234567,
+                "checksum": "dde60621880aa996c42e356a687744ef",
+                "checksum_type": 0,
+                "compression_type": None,
+                "extra": {"image": {"arch": "x86_64"}},
+                "filename": file_name,
+                "id": 234792,
+                "metadata_only": False,
+                "size": 16077,
+                "type_description": "JSON data",
+                "type_extensions": "json",
+                "type_id": 49,
+                "type_name": type_name,
+            }
+        )
+
+    name = "rhcos"
+    version = "414.92"
+    release = "1"
+    nvr = f"{name}-{version}-{release}"
+
+    source = Source.get(
+        f"koji:https://koji.example.com/?vmi_build={nvr}", basedir=koji_dir
+    )
+    assert not fake_koji.last_url
+
+    fake_koji.build_data[1234] = {
+        "id": 1234,
+        "name": name,
+        "version": version,
+        "release": release,
+        "nvr": nvr,
+        "completion_time": "2022-12-20 12:30:25",
+        "extra": {"typeinfo": {"image": {"arch": "x86_64"}}},
+    }
+
+    fake_koji.build_data[nvr] = fake_koji.build_data[1234]
+
+    fake_koji.insert_archives(archives=archives, build_nvr=nvr)
+
+    path = "/path/to/rhcos/metadata"
+    meta_data = """{    
+    "name": "rhcos",
+    "summary": "OpenShift 4",
+    "images": {
+        "azure": {
+            "path": "rhcos-414.92-0-azure.x86_64.vhd.gz",
+            "sha256": "2cd817331af29093e2eaa025139ebddd6008c193970b06b35afbbdbebae0ce3e",
+            "dest": ["https://example.windows.net/imagebucket/rhcos-414.92-0-azure.x86_64.vhd"]},
+  "aws": {
+        "path": "rhcos-414.92-0-aws.x86_64.vmdk.gz",
+        "sha256": "4ef7806152bd89ce44326ff746c4f1883ad543885e980bce59821df2d946ea4c"
+        }},
+ "azure": {
+        "image": "rhcos-414.92-0-azure.x86_64.vhd",
+        "url": "https://example.windows.net/imagebucket/rhcos-414.92-0-azure.x86_64.vhd"},
+ "amis": [
+    {"name": "us-gov-west-1",
+   "hvm": "ami-01"
+   },
+    {"name": "us-east-1",
+   "hvm": "ami-02"
+   }]}"""
+
+    patch.object(os.path, "join", return_value=path)
```

Comment:
This mocking doesn't seem like it should be necessary, because you can easily point the koji source to point at any directory you want.  In fact you're already pointing it at `koji_dir`, so you just need to write the meta.json file to the correct path under that dir and it should work without mocking.

Can you please do it that way?  This way is really nasty, because:

1) You're not checking that 'join' is used correctly at all.  The test can pass even if the code is totally broken and calls 'join' with incorrect arguments.

2) Mocking like this requires more maintenance in the log term, because perfectly correct changes in the code can cause the test to break just by not calling exactly the methods that you've mocked here.

Reply:
I have removed it, PTAL